### PR TITLE
Nix channel pins and improvements

### PR DESCRIPTION
**Pull request by piegamesde**
Closes #22

### CODE REVIEW

File: src/channel.rs [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+impl Updatable for Pin {
+    type Version = ChannelVersion;
+    type Hashes = ChannelHash;
+
+    async fn update(&self, _old: Option<&ChannelVersion>) -> Result<ChannelVersion> {
+        /* We want to get from something like https://channels.nixos.org/nixos-21.11
+         * to https://releases.nixos.org/nixos/21.11/nixos-21.11.335807.df4f1f7cc3f/nixexprs.tar.xz
+         */
+        let mut url = build_client()?
+            .get(&format!("https://channels.nixos.org/{}", self.name))
```
Comment (andir) [inline src/channel.rs:44]: Lets just use `https://channels.nixos.org/<name>/nixexpr.tar.xz` from the beginning. That removes the needs for the `TODO error handling` below and is also compatible with situation in which there might not be a redirect.

File: src/cli.rs [hunk 2]
```diff
@@ -450,18 +505,25 @@ impl Opts {
     async fn update(&self, opts: &UpdateOpts) -> Result<()> {
         let mut pins = self.read_pins()?;
 
+        let strategy = match (opts.partial, opts.full) {
+            (false, false) => UpdateStrategy::Normal,
+            (false, true) => UpdateStrategy::Full,
+            (true, false) => UpdateStrategy::HashesOnly,
+            (true, true) => panic!("partial and full are mutually exclusive"),
```
Comment (andir) [inline src/cli.rs:512]: Panic isn't really a great way of handling this. Can we properly log this with an exit code of 1?

Comment (piegamesde) [reply src/cli.rs:512]: The user can't specify both; the CLI arument parse will take care of this: "error: The argument '--partial' cannot be used with '--full'". On the other hand, programmer mistakes should fail loudly.

**Comment by andir** [comment]
```
[nix-shell:/tmp/demo]$ npins --help
npins 0.1.0

USAGE:
    npins [OPTIONS] <SUBCOMMAND>

FLAGS:
    -h, --help       Prints help information
    -V, --version    Prints version information

OPTIONS:
    -d, --directory <folder>    Base folder for sources.json and the boilerplate default.nix [env: NPINS_DIRECTORY=]
                                [default: npins]

SUBCOMMANDS:
    add           Adds a new pin entry
    fetch         Query some release information and then print out the entry
    help          Prints this message or the help of the given subcommand(s)
    import-niv    Try to import entries from Niv
    init          Intializes the npins directory. Running this multiple times will restore/upgrade the `default.nix`
                  and never touch your sources.json
    remove        Removes one pin entry
    show          Lists the current pin entries
    update        Updates all or the given pin to the latest version
    upgrade       Upgrade the sources.json and default.nix to the latest format version. This may occasionally break
                  Nix evaluation!

[nix-shell:/tmp/demo]$ npins init --help
npins-init 0.1.0
Intializes the npins directory. Running this multiple times will restore/upgrade the `default.nix` and never touch your
sources.json

USAGE:
    npins init [FLAGS] [OPTIONS]

FLAGS:
        --bare    Don't add an initial `nixpkgs` entry
    -h, --help    Prints help information

OPTIONS:
    -d, --directory <folder>    Base folder for sources.json and the boilerplate default.nix [env: NPINS_DIRECTORY=]
                                [default: npins]

[nix-shell:/tmp/demo]$ npins init --bare
[INFO ] Welcome to npins!
[INFO ] Creating `npins` directory
[INFO ] Writing default.nix
[INFO ] Writing initial sources.json (empty)
[INFO ] Done.

[nix-shell:/tmp/demo]$ npins add channel nixos-21.11

[nix-shell:/tmp/demo]$ nix repl
warning: unknown setting 'experimental-features'
Welcome to Nix version 2.3.16. Type :? for help.

nix-repl> pins = import ./npins

nix-repl> pkgs = import pins."nixos-21.11" { }

[20.4 MiB DL]
error: hash mismatch in file downloaded from 'https://releases.nixos.org/nixos/21.11/nixos-21.11.335820.521e4d7d13b/nixexprs.tar.xz':
  wanted: sha256:16p4cxpzlwdj6pv29a2dd0cv439v0dwlddzkdh1wnv7075nw2lyh
  got:    sha256:1d79qhyqa1h58y6k7gacm72is8hpla8db3sbqpkq6g813r8sdlym

```

Two remarks:
1. The `npins add channel ... ` command doesn't log anything but hangs there for a few seconds.
2. The hash doesn't match :D

**Comment by piegamesde** [comment]
1. Yes, I moved the logging commit to the other PR
2. I cannot reproduce the hash mismatch. I tried it twice and it works for me :D

**Comment by mweinelt** [comment]
```
❯ nix-prefetch-url --unpack https://releases.nixos.org/nixos/21.11/nixos-21.11.335820.521e4d7d13b/nixexprs.tar.xz
path is '/nix/store/01j26k9mn7kyk5k16qjjigw3gwsi86az-nixexprs.tar.xz'
16p4cxpzlwdj6pv29a2dd0cv439v0dwlddzkdh1wnv7075nw2lyh
```

File: src/cli.rs (hunk 2 above)
Comment (piegamesde) [reply src/cli.rs:512]: Moved to #28

**Pull request merged**